File: PHP/135. Candy.php
<?php
/**
  * 135. Candy
  * There are n children standing in a line. Each child is assigned a rating value given in the integer array ratings.
  * You are giving candies to these children subjected to the following requirements:
  * Each child must have at least one candy.
  * Children with a higher rating get more candies than their neighbors.
  * Return the minimum number of candies you need to have to distribute the candies to the children.
  **/
class Solution {

    /**
     * @param Integer[] $ratings
     * @return Integer
     */
    function candy($ratings) {
        $kount = count($ratings);
        // Every kid gets at least 1 candy.
        $candies = array_fill(0, $kount, 1);
        // Forwards.
        for ($i = 1; $i < $kount; $i++) {
            // If the current kid is higher than the left kid, current kid gets one more than left kid.
            if ($ratings[$i] > $ratings[$i-1]) { 
                $candies[$i] = $candies[$i - 1] + 1; 
            }
        }
        // Backwards.
        for ($i = $kount - 2; $i >= 0; $i--) {
            // If left kid is higher than right kid, make sure that left kid gets at least 1 more than right kid.
            if ($ratings[$i] > $ratings[$i + 1]) {
                $candies[$i] = max($candies[$i], ($candies[$i + 1] + 1));
            }
        }
        return array_sum($candies);
    }
}

$c = new Solution;
print_r($c->candy([1,0,2]));
?>

File: Swift/135. Candy.swift
/*
There are n children standing in a line. Each child is assigned a rating value given in the integer array ratings.
You are giving candies to these children subjected to the following requirements:
Each child must have at least one candy.
Children with a higher rating get more candies than their neighbors.
Return the minimum number of candies you need to have to distribute the candies to the children.
    
135. Candy    
*/
class Solution {
    func candy(_ ratings: [Int]) -> Int {
        let n = ratings.count
        var candies = [Int](repeating: 1, count: n)
        for i in 1..<n {
            if ratings[i] > ratings[i - 1] {
                candies[i] = candies[i - 1] + 1
            }
        }
        for i in stride(from: n - 2, through: 0, by: -1) {
            if ratings[i] > ratings[i + 1] {
                candies[i] = max(candies[i], (candies[i + 1] + 1))
            }
        }
        return candies.reduce(0, +)
    }
}
let solution = Solution()
print(solution.candy([1,0,2])) // Expect 5
print(solution.candy([1,2,2]))  // Expect 4